File: python/leetcode/leet_1481.py
# Least Number of Unique Integers after K Removals
# https://leetcode.com/problems/least-number-of-unique-integers-after-k-removals/description/?envType=daily-question&envId=2024-02-16

from collections import Counter


def find_least_num_of_unique_ints(arr, k) -> int:
    counter = Counter(arr)
    sorted_counts = sorted(counter.values())
    removed_elements = 0

    for i in range(len(sorted_counts)):
        removed_elements += sorted_counts[i]
        if removed_elements > k:
            return len(sorted_counts) - i

    return 0

File: rust/leet_code/src/leet_1481.rs
// Least Number of Unique Integers after K Removals
// https://leetcode.com/problems/least-number-of-unique-integers-after-k-removals/?envType=daily-question&envId=2024-02-16

use counter::Counter;
use std::collections::HashMap;

fn find_least_num_of_unique_ints(arr: Vec<i32>, k: i32) -> i32 {
    let counter = arr.into_iter().collect::<Counter<_>>();

    let mut sorted_counts: Vec<_> = counter.iter().collect();
    sorted_counts.sort_by_key(|&(_, count)| count);

    let mut removed_elements = 0;

    for (_, count) in sorted_counts.iter() {
        removed_elements += *count;
        if removed_elements > k {
            return sorted_counts.len() as i32 - removed_elements + *count;
        }
    }
    0
}
